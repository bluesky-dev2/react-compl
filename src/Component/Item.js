import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import cssClasses from './../cssClasses';
import cx from 'classnames';

class Item extends PureComponent {
  static propTypes = {
    label: PropTypes.node.isRequired,
    icon: PropTypes.string,
    leftIcon: PropTypes.string,
    rightIcon: PropTypes.string,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
    data: PropTypes.any,
    target: PropTypes.any
  };

  static defaultProps = {
    icon: '',
    leftIcon: '',
    rightIcon: '',
    disabled: false,
    onClick: () => {}
  };

  handleClick = () => this.props.onClick(this, this.props.target);

  buildItem() {
    return (
      <div className={cssClasses.ITEM_DATA}>
        {this.hasIcon()}
        {this.props.label}
      </div>
    );
  }

  hasIcon() {
    return this.props.icon.length > 0
            ? <span className={`${cssClasses.ITEM_ICON} ${this.props.icon}`} />
            : '';
  }

  render() {
    const className = cx(cssClasses.ITEM, {
      [`${cssClasses.ITEM_DISABLED}`]: this.props.disabled
    });
    return (
      <div className={className} onClick={this.handleClick}>
        {this.buildItem()}
      </div>
    );
  }
}

export default Item;
