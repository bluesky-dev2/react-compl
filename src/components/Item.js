import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import styles from './styles';

class Item extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    style: PropTypes.object,
    nativeEvent: PropTypes.object,
    disabled: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
    onClick: PropTypes.func,
    data: PropTypes.any,
    dataFromProvider: PropTypes.any,
    refsFromProvider: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.object)
    ])
  };

  static defaultProps = {
    className: null,
    style: {},
    disabled: false,
    onClick: () => {},
    nativeEvent: {},
    data: null,
    refsFromProvider: [],
    dataFromProvider: null
  };

  disabled = false;

  handleClick = e => {
    this.disabled
      ? e.stopPropagation()
      : this.props.onClick({
          event: this.props.nativeEvent,
          ref: this.props.refsFromProvider,
          data: this.props.data,
          dataFromProvider: this.props.dataFromProvider
        });
  };

  render() {
    const {
      className,
      disabled,
      style,
      children,
      data,
      refsFromProvider,
      dataFromProvider,
      nativeEvent
    } = this.props;
    
    this.disabled =
      typeof disabled === 'function'
        ? disabled({
            event: nativeEvent,
            ref: refsFromProvider,
            data: data,
            dataFromProvider: dataFromProvider
          })
        : disabled;

    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]: this.disabled
    });

    return (
      <div
        className={cssClasses}
        style={style}
        onClick={this.handleClick}
        role="presentation"
      >
        <div className={styles.itemContent}>{children}</div>
      </div>
    );
  }
}

export default Item;
