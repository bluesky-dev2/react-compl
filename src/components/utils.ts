import { Children, cloneElement, ReactNode, ReactElement } from 'react';

import { BooleanPredicate, PredicateParams, TriggerEvent } from '../types';

export function isFn(v: any): v is Function {
  return typeof v === 'function';
}

export function isStr(v: any): v is String {
  return typeof v === 'string';
}

export function cloneItems(
  children: ReactNode,
  props: { triggerEvent: TriggerEvent; propsFromTrigger?: object }
) {
  return Children.map(
    // remove null item
    Children.toArray(children).filter(Boolean),
    (item) => cloneElement(item as ReactElement<any>, props)
  );
}

export function getMousePosition(e: TriggerEvent) {
  const pos = {
    x: (e as MouseEvent).clientX,
    y: (e as MouseEvent).clientY,
  };

  const touch = (e as TouchEvent).changedTouches;

  if (touch) {
    pos.x = touch[0].clientX;
    pos.y = touch[0].clientY;
  }

  if (!pos.x || pos.x < 0) pos.x = 0;

  if (!pos.y || pos.y < 0) pos.y = 0;

  return pos;
}

export function getPredicateValue(
  predicate: BooleanPredicate,
  payload: PredicateParams
) {
  return isFn(predicate) ? predicate(payload) : predicate;
}
