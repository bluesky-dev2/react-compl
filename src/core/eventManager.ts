export type EventType = string | number | symbol;
export type Handler<T = any> = (args: T) => void;

export interface EventManager<E = EventType> {
  on<T = any>(event: E, handler: Handler<T>): EventManager<E>;
  off<T = any>(event: E, handler?: Handler<T>): EventManager<E>;
  emit<T = any>(event: E, args?: T): EventManager<E>;
}

function createEventManager<E = EventType>(): EventManager<E> {
  const eventList = new Map<E, Set<Handler>>();

  return {
    on<T = any>(event: E, handler: Handler<T>) {
      eventList.has(event)
        ? eventList.get(event)!.add(handler)
        : eventList.set(event, new Set([handler]));
      return this;
    },
    off<T = any>(event: E, handler: Handler<T>) {
      eventList.has(event) && eventList.get(event)!.delete(handler);
      return this;
    },
    emit<T = any>(event: E, args: T) {
      eventList.has(event) &&
        eventList.get(event)!.forEach((handler: Handler<T>) => {
          handler(args);
        });
      return this;
    },
  };
}

export const eventManager = createEventManager();
