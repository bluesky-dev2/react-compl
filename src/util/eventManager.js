const eventManager = {
  eventList: new Map(),
  on(event, callback) {
    this.eventList.has(event) || this.eventList.set(event, new Set());
    this.eventList.get(event).add(callback);

    return () => this.eventList.get(event).delete(callback);
  },
  emit(event, ...args) {
    if (!this.eventList.has(event)) {
      console.warn(
        `<${event}> Event is not registered. Did you forgot to bind the event ?`
      );
      return false;
    }
    this.eventList.get(event).forEach(cb => cb.call(this, ...args));

    return true;
  }
};

export default eventManager;
