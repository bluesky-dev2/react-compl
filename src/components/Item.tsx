import React, { ReactNode } from 'react';
import cx from 'clsx';

import {
  ItemParams,
  InternalProps,
  BooleanPredicate,
  HandlerParamsEvent,
} from '../types';
import { useRefTrackerContext } from './RefTrackerProvider';
import { NOOP, STYLE } from '../constants';
import { getPredicateValue } from './utils';

export interface ItemProps
  extends InternalProps,
    Omit<React.HTMLAttributes<HTMLElement>, 'hidden' | 'disabled' | 'onClick'> {
  /**
   * Any valid node that can be rendered
   */
  children: ReactNode;

  /**
   * Passed to the `Item` onClick callback. Accessible via `data`
   */
  data?: any;

  /**
   * Disable `Item`. If a function is used, a boolean must be returned
   *
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   *
   * ```
   * function isItemDisabled({ triggerEvent, props, data }: PredicateParams<type of props, type of data>): boolean
   * <Item disabled={isItemDisabled} data={data}>content</Item>
   * ```
   */
  disabled?: BooleanPredicate;

  /**
   * Hide the `Item`. If a function is used, a boolean must be returned
   *
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   *
   * ```
   * function isItemHidden({ triggerEvent, props, data }: PredicateParams<type of props, type of data>): boolean
   * <Item hidden={isItemHidden} data={data}>content</Item>
   * ```
   */
  hidden?: BooleanPredicate;

  /**
   * Callback when the `Item` is clicked.
   *
   * @param event The event that occured on the Item node
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   * ```
   * function handleItemClick({ triggerEvent, event, props, data }: ItemParams<type of props, type of data>){
   *    // retrieve the id of the Item or any other dom attribute
   *    const id = e.currentTarget.id;
   *
   *    // access the props and the data
   *    console.log(props, data);
   *
   *    // access the coordinate of the mouse when the menu has been displayed
   *    const {  clientX, clientY } = triggerEvent;
   *
   * }
   *
   * <Item id="item-id" onClick={handleItemClick} data={{key: 'value'}}>Something</Item>
   * ```
   */
  onClick?: (args: ItemParams) => void;
}

export const Item: React.FC<ItemProps> = ({
  children,
  className,
  style,
  triggerEvent,
  data,
  propsFromTrigger,
  onClick = NOOP,
  disabled = false,
  hidden = false,
  ...rest
}) => {
  const refTracker = useRefTrackerContext();
  const handlerParams = {
    data,
    triggerEvent: triggerEvent as HandlerParamsEvent,
    props: propsFromTrigger,
  };
  const isDisabled = getPredicateValue(disabled, handlerParams);
  const isHidden = getPredicateValue(hidden, handlerParams);

  function handleClick(e: React.MouseEvent<HTMLElement>) {
    (handlerParams as ItemParams).event = e;
    isDisabled ? e.stopPropagation() : onClick(handlerParams as ItemParams);
  }

  function trackRef(node: HTMLElement | null) {
    if (node && !isDisabled)
      refTracker.set(node, {
        node,
        isSubmenu: false,
      });
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLElement>) {
    if (e.key === 'Enter') {
      (handlerParams as ItemParams).event = e;
      onClick(handlerParams as ItemParams);
    }
  }

  if (isHidden) return null;

  const cssClasses = cx(STYLE.item, className, {
    [`${STYLE.itemDisabled}`]: isDisabled,
  });

  return (
    <div
      {...rest}
      className={cssClasses}
      style={style}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      ref={trackRef}
      tabIndex={-1}
      role="menuitem"
      aria-disabled={isDisabled}
    >
      <div className={STYLE.itemContent}>{children}</div>
    </div>
  );
};
