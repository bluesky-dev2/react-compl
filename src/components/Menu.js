/* global: window */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import cloneItem from './cloneItem';

import { HIDE_ALL, DISPLAY_MENU } from '../utils/actions';
import styles from '../utils/styles';
import eventManager from '../utils/eventManager';

const KEY = {
  ENTER: 13,
  ESC: 27,
  ARROW_UP: 38,
  ARROW_DOWN: 40,
  ARROW_LEFT: 37,
  ARROW_RIGHT: 39
};

class Menu extends Component {
  static propTypes = {
    /**
     * Unique id to identify the menu
     */
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,

    /**
     * Any valid node that can be rendered
     */
    children: PropTypes.node.isRequired,
    /**
     * Append given css classes
     */
    className: PropTypes.string,
    /**
     * Append given inline style
     */
    style: PropTypes.object,
    /**
     * Theme is appended to `react-contexify__theme--${given theme}`.
     * 
     * Built-in theme are `light` and `dark`
     */
    theme: PropTypes.string,

    /**
     * Animation is appended to `.react-contexify__will-enter--${given animation}`
     * 
     * Built-in animations are fade, flip, pop, zoom
     */
    animation: PropTypes.string
  };

  static defaultProps = {
    className: null,
    style: {},
    theme: null,
    animation: null
  };

  state = {
    x: 0,
    y: 0,
    visible: false,
    nativeEvent: null,
    propsFromTrigger: {}
  };

  menuRef = null;
  unsub = [];

  componentDidMount() {
    this.unsub.push(eventManager.on(DISPLAY_MENU(this.props.id), this.show));
    this.unsub.push(eventManager.on(HIDE_ALL, this.hide));
  }

  componentWillUnmount() {
    this.unsub.forEach(cb => cb());
    this.unBindWindowEvent();
  }

  bindWindowEvent = () => {
    window.addEventListener('resize', this.hide);
    window.addEventListener('contextmenu', this.hide);
    window.addEventListener('mousedown', this.hide);
    window.addEventListener('click', this.hide);
    window.addEventListener('scroll', this.hide);
    window.addEventListener('keydown', this.handleKeyboard);
  };

  unBindWindowEvent = () => {
    window.removeEventListener('resize', this.hide);
    window.removeEventListener('contextmenu', this.hide);
    window.removeEventListener('mousedown', this.hide);
    window.removeEventListener('click', this.hide);
    window.removeEventListener('scroll', this.hide);
    window.removeEventListener('keydown', this.handleKeyboard);
  };

  onMouseEnter = () => window.removeEventListener('mousedown', this.hide);

  onMouseLeave = () => window.addEventListener('mousedown', this.hide);

  hide = e => {
    // Safari trigger a click event when you ctrl + trackpad
    // Firefox:  trigger a click event when right click occur
    if (
      typeof e !== 'undefined' &&
      (e.button === 2 || e.ctrlKey === true) &&
      e.type !== 'contextmenu'
    ) {
      return;
    }

    this.unBindWindowEvent();
    this.setState({ visible: false });
  };

  handleKeyboard = e => {
    if (e.keyCode === KEY.ENTER || e.keyCode === KEY.ESC) {
      this.unBindWindowEvent();
      this.setState({ visible: false });
    }
  };

  setRef = ref => {
    this.menuRef = ref;
  };

  setMenuPosition() {
    const { innerWidth: windowWidth, innerHeight: windowHeight } = window;
    const { offsetWidth: menuWidth, offsetHeight: menuHeight } = this.menuRef;
    let { x, y } = this.state;

    if (x + menuWidth > windowWidth) {
      x -= x + menuWidth - windowWidth;
    }

    if (y + menuHeight > windowHeight) {
      y -= y + menuHeight - windowHeight;
    }

    this.setState(
      {
        x,
        y
      },
      this.bindWindowEvent
    );
  }

  getMousePosition(e) {
    const pos = {
      x: e.clientX,
      y: e.clientY
    };

    if (
      e.type === 'touchend' &&
      (pos.x === null || pos.y === null) &&
      (e.changedTouches !== null && e.changedTouches.length > 0)
    ) {
      pos.x = e.changedTouches[0].clientX;
      pos.y = e.changedTouches[0].clientY;
    }

    if (pos.x === null || pos.x < 0) {
      pos.x = 0;
    }

    if (pos.y === null || pos.y < 0) {
      pos.y = 0;
    }

    return pos;
  }

  show = (e, props) => {
    e.stopPropagation();
    eventManager.emit(HIDE_ALL);

    const { x, y } = this.getMousePosition(e);

    this.setState(
      {
        visible: true,
        x,
        y,
        nativeEvent: e,
        propsFromTrigger: props
      },
      this.setMenuPosition
    );
  };

  render() {
    const { theme, animation, style, className, children } = this.props;
    const { visible, nativeEvent, propsFromTrigger, x, y } = this.state;

    const cssClasses = cx(styles.menu, className, {
      [styles.theme + theme]: theme !== null,
      [styles.animationWillEnter + animation]: animation !== null
    });
    const menuStyle = {
      ...style,
      left: x,
      top: y + 1,
      opacity: 1
    };

    return (
      visible && (
        <div
          className={cssClasses}
          style={menuStyle}
          ref={this.setRef}
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
        >
          <div>
            {cloneItem(children, { nativeEvent, propsFromTrigger })}
          </div>
        </div>
      )
    );
  }
}

Menu.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  foo: PropTypes.string,
  bar: PropTypes.object,
};
Menu.defaultProps = {
  foo: 'abc',
  bar: 'xyz'
}



export default Menu;
