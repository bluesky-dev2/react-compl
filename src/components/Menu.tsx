/* global: window */
import React, {
  ReactNode,
  useEffect,
  useReducer,
  useRef,
  useState,
} from 'react';
import cx from 'clsx';

import { Portal, PortalProps } from './Portal';
import { RefTrackerProvider } from './RefTrackerProvider';

import { eventManager } from '../core/eventManager';
import { MouseOrTouchEvent, MenuId, ContextMenuParams } from '../types';
import { usePrevious, useRefTracker } from '../hooks';
import { createMenuController } from './menuController';
import { NOOP, STYLE, EVENT } from '../constants';
import { cloneItems, getMousePosition } from './utils';

export interface MenuProps
  extends PortalProps,
    Omit<React.HTMLAttributes<HTMLElement>, 'id'> {
  /**
   * Unique id to identify the menu. Use to Trigger the corresponding menu
   */
  id: MenuId;

  /**
   * Any valid node that can be rendered
   */
  children: ReactNode;

  /**
   * Theme is appended to `react-contexify__theme--${given theme}`.
   *
   * Built-in theme are `light` and `dark`
   */
  theme?: string;

  /**
   * Animation is appended to `.react-contexify__will-enter--${given animation}`
   *
   * Built-in animations are fade, flip, pop, zoom
   */
  animation?: string;

  /**
   * Invoked when the menu is shown.
   */
  onShown?: () => void;

  /**
   * Invoked when the menu is hidden.
   */
  onHidden?: () => void;
}

interface MenuState {
  x: number;
  y: number;
  visible: boolean;
  nativeEvent: MouseOrTouchEvent;
  propsFromTrigger: any;
}

function reducer(state: MenuState, payload: Partial<MenuState>) {
  return { ...state, ...payload };
}

export const Menu: React.FC<MenuProps> = ({
  id,
  theme,
  animation,
  style,
  className,
  children,
  mountNode,
  onHidden = NOOP,
  onShown = NOOP,
  ...rest
}) => {
  const [state, setState] = useReducer(reducer, {
    x: 0,
    y: 0,
    visible: false,
    nativeEvent: {} as MouseOrTouchEvent,
    propsFromTrigger: null,
  });
  const nodeRef = useRef<HTMLDivElement>(null);
  const didMount = useRef(false);
  const wasVisible = usePrevious(state.visible);
  const refTracker = useRefTracker();
  const [menuController] = useState(() => createMenuController());

  // subscribe event manager
  useEffect(() => {
    didMount.current = true;
    eventManager.on(id, show).on(EVENT.HIDE_ALL, hide);

    return () => {
      eventManager.off(id, show).off(EVENT.HIDE_ALL, hide);
    };
  }, [id]);

  // handle show/ hide callback
  useEffect(() => {
    if (didMount.current && state.visible !== wasVisible) {
      state.visible ? onShown() : onHidden();
    }
  }, [state.visible, onHidden, onShown]);

  // collect menu items for keyboard navigation
  useEffect(() => {
    if (!state.visible) {
      refTracker.clear();
    } else {
      menuController.init(Array.from(refTracker.values()));
    }
  }, [state.visible, menuController, refTracker]);

  // compute menu position
  useEffect(() => {
    if (state.visible) {
      const { innerWidth: windowWidth, innerHeight: windowHeight } = window;
      const {
        offsetWidth: menuWidth,
        offsetHeight: menuHeight,
      } = nodeRef.current!;
      let { x, y } = state;

      if (x + menuWidth > windowWidth) {
        x -= x + menuWidth - windowWidth;
      }

      if (y + menuHeight > windowHeight) {
        y -= y + menuHeight - windowHeight;
      }

      setState({
        x,
        y,
      });
    }
  }, [state.visible]);

  // subscribe dom events
  useEffect(() => {
    function handleKeyboard(e: KeyboardEvent) {
      switch (e.key) {
        case 'Enter':
          if (!menuController.openSubmenu()) hide();
          break;
        case 'Escape':
          hide();
          break;
        case 'ArrowUp':
          menuController.moveUp();
          break;
        case 'ArrowDown':
          menuController.moveDown();
          break;
        case 'ArrowRight':
          menuController.openSubmenu();
          break;
        case 'ArrowLeft':
          menuController.closeSubmenu();
          break;
      }
    }

    if (state.visible) {
      window.addEventListener('resize', hide);
      window.addEventListener('contextmenu', hide);
      window.addEventListener('click', hide);
      window.addEventListener('scroll', hide);
      window.addEventListener('keydown', handleKeyboard);

      // This let us debug the menu in the console in dev mode
      if (process.env.NODE_ENV !== 'development') {
        window.addEventListener('blur', hide);
      }
    }

    return () => {
      window.removeEventListener('resize', hide);
      window.removeEventListener('contextmenu', hide);
      window.removeEventListener('click', hide);
      window.removeEventListener('scroll', hide);
      window.removeEventListener('keydown', handleKeyboard);

      if (process.env.NODE_ENV !== 'development') {
        window.removeEventListener('blur', hide);
      }
    };
  }, [state.visible, menuController]);

  function show({ event, props, position }: ContextMenuParams) {
    event.stopPropagation();
    const { x, y } = position || getMousePosition(event);

    // prevent react from batching the state update
    // if the menu is already visible we have to recompute bounding rect based on position
    setTimeout(() => {
      setState({
        visible: true,
        x,
        y,
        nativeEvent: event,
        propsFromTrigger: props,
      });
    }, 0);
  }

  function hide(event?: Event) {
    // Safari trigger a click event when you ctrl + trackpad
    // Firefox:  trigger a click event when right click occur
    const e = event as KeyboardEvent & MouseEvent;

    if (
      typeof e !== 'undefined' &&
      (e.button === 2 || e.ctrlKey === true) &&
      e.type !== 'contextmenu'
    ) {
      return;
    }

    setState({ visible: false });
  }

  const cssClasses = cx(STYLE.menu, className, {
    [STYLE.theme + theme]: theme,
    [STYLE.animationWillEnter + animation]: animation,
  });
  const { visible, nativeEvent, propsFromTrigger, x, y } = state;
  const menuStyle = {
    ...style,
    left: x,
    top: y + 1,
    opacity: 1,
  };

  return (
    <Portal mountNode={mountNode}>
      <RefTrackerProvider refTracker={refTracker}>
        {visible && (
          <div
            {...rest}
            className={cssClasses}
            style={menuStyle}
            ref={nodeRef}
            role="menu"
          >
            {cloneItems(children, {
              propsFromTrigger,
              nativeEvent: nativeEvent,
            })}
          </div>
        )}
      </RefTrackerProvider>
    </Portal>
  );
};
