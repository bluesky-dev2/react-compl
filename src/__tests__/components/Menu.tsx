import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Menu } from '../../components/Menu';
import { Item } from '../../components/Item';
import { eventManager } from '../../utils/eventManager';
import { HIDE_ALL, DISPLAY_MENU } from '../../utils/actions';

beforeEach(() => eventManager.eventList.clear());

const menuId = 'foo';

describe('Menu', () => {
  it('Should bind event when component did mount and unbind all the event related to the component when will unmount', () => {
    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    expect(eventManager.eventList.get(DISPLAY_MENU(menuId))!.size).toBe(1);
    expect(eventManager.eventList.get(HIDE_ALL)!.size).toBe(1);
    component.unmount();
    expect(eventManager.eventList.get(DISPLAY_MENU(menuId))!.size).toBe(0);
    expect(eventManager.eventList.get(HIDE_ALL)!.size).toBe(0);
  });

  it('Should render null if the context menu is not visible', () => {
    const component = shallow(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    component.setState({ visible: true });
    expect(component.html()).not.toBeNull();

    component.setState({ visible: false });
    expect(component.html()).toBeNull();
  });

  it('Should hide Menu when `hide` method is called', () => {
    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );

    component.setState({ visible: true });
    const instance = component.instance() as Menu;
    instance.hide();

    expect(component.state('visible')).toBe(false);
    expect(component.html()).toBeNull();
  });

  it('Should remove null Item from menu', () => {
    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
        {null}
      </Menu>
    );
    component.setState({ visible: true });
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should remove the mousedown event on window object when mouse enter the context menu', () => {
    window.removeEventListener = jest.fn();

    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );

    component.setState({ visible: true });
    component.simulate('mouseenter');
    expect(window.removeEventListener).toHaveBeenCalled();
  });

  it('Should add the mousedown event on window object when mouse leave the context menu', () => {
    window.addEventListener = jest.fn();

    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );

    component.setState({ visible: true });
    component.simulate('mouseleave');
    expect(window.addEventListener).toHaveBeenCalled();
  });

  it('Should display the menu when corresponding event is emitted', () => {
    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    expect(component.html()).toBeNull();

    eventManager.emit(
      DISPLAY_MENU(menuId),
      { stopPropagation() {}, clientX: 1, clientY: 1 },
      {}
    );

    expect(component.html()).not.toBeNull();
  });

  it('Should be able to handle MouseEvent and TouchEvent', () => {
    let position = {} as {
      x: number;
      y: number;
    };

    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    const instance = component.instance() as Menu;

    const mouseEvent = new MouseEvent('click', {
      clientX: 10,
      clientY: 10
    });

    const touchInit = {
      clientX: 12,
      clientY: 12
    } as Touch;

    const touchEvent = new TouchEvent('touchend', {
      changedTouches: [touchInit]
    });

    position = instance.getMousePosition(mouseEvent);
    expect(position).toMatchObject({ x: 10, y: 10 });

    position = instance.getMousePosition(touchEvent);
    expect(position).toMatchObject({ x: 12, y: 12 });
  });

  it("Should set a default position if not able to determine one", () => {
    let position = {} as {
      x: number;
      y: number;
    };

    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    const instance = component.instance() as Menu;

    position = instance.getMousePosition(new MouseEvent('click', {
      clientX: -1,
      clientY: -1
    }));
    expect(position).toMatchObject({ x: 0, y: 0 });

    position = instance.getMousePosition(new MouseEvent('click', {
      clientX: undefined,
      clientY: undefined
    }));
    expect(position).toMatchObject({ x: 0, y: 0 });
  });

  it('Should have the same behavior accross different browser', () => {
    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );

    expect(component.html()).toBeNull();

    eventManager.emit(
      DISPLAY_MENU(menuId),
      { stopPropagation() {}, clientX: 1, clientY: 1 },
      {}
    );

    expect(component.state('visible')).toBe(true);

    // Test for Firefox
    eventManager.emit(HIDE_ALL, {
      button: 2,
      type: 'click'
    });

    expect(component.state('visible')).toBe(true);

    // Test for Safari
    eventManager.emit(HIDE_ALL, {
      ctrlKey: true,
      type: 'click'
    });

    expect(component.state('visible')).toBe(true);
  });

  it('Should hide menu when `enter` or `escape` is pressed down', () => {
    const windowEvent: {
      [key: string]: (arg: any) => any;
    } = {};
    // mock to simulate keyboard events
    window.addEventListener = jest.fn((event, cb) => {
      windowEvent[event] = cb;
    });

    const component = mount(
      <Menu id={menuId}>
        <Item>bar</Item>
      </Menu>
    );
    component.setState({ visible: true });
    const instance = component.instance() as Menu;
    instance.bindWindowEvent();

    windowEvent.keydown({ keyCode: 13 });
    expect(component.state('visible')).toBe(false);

    component.setState({ visible: true });

    windowEvent.keydown({ keyCode: 27 });
    expect(component.state('visible')).toBe(false);
  });
});
