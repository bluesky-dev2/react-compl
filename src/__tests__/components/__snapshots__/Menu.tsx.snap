// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Should remove null Item from menu 1`] = `
<Menu
  id="foo"
>
  <div
    className="react-contexify"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "left": 0,
        "opacity": 1,
        "top": 1,
      }
    }
  >
    <div>
      <Item
        disabled={false}
        key=".$.0"
        nativeEvent={Object {}}
        onClick={[Function]}
        propsFromTrigger={Object {}}
      >
        <div
          className="react-contexify__item"
          onClick={[Function]}
          role="presentation"
        >
          <div
            className="react-contexify__item__data"
          >
            bar
          </div>
        </div>
      </Item>
    </div>
  </div>
</Menu>
`;
